Search 자동완성창 UI 작성

sangwon21 opened the pull request:
- what
  - 검색어를 입력하면 해당 검색어로 시작하는 단어 목록이 화면에 등장하게끔 했습니다.
  - 자동완성 결과는 키보드 위/아래키로 이동할 수 있게 했습니다.
  - 노출된 데이터 중 검색어와 일치하는 단어는 색깔이 하이라이트 되여 보여지게 했습니다.
  - 검색 대상이 되는 단어들은 영문학 고전과 꼭 봐야할 영어권 영화 100선을 선정했습니다.

- how
  - 단어 목록이 화면에 렌더링 되기 위해서는 먼저, 데이터가 필요합니다. 그래서 데이터를 받아오기 위하여 서버에 fetch하는 작업을 시작했습니다.
  - 자동완성 결과를 키보드 위 아래로 움직이는 건, 키보드 이벤트를 이용하여 해당 키가 들어왔을 때, 알맞는 리스너 함수(listener)를 작성하여 해결했습니다.
  - 일치하는 단어의 색깔이 하이라이트 되는 부분은 css의 클래스를 줌으로써 해결했습니다.

- why
  - 데모를 보여주기 위해서, 로컬 서버를 이용할 수 없어, firebase를 이용하여 배포했습니다.

- 개선할 점
  - 로컬 서버를 이용해서 개발할 때, 고려하지 않았던 문제점, 응답 속도를 해결하는 방법을 찾아야 합니다.
  - 특히, 데모 주소에 처음 들어갔을 때, 검색창이 보이지 않고 새로고침을 해야만 보이는 오류가 있는데 이를 해결해야 합니다. 
  - 동작 중에 사용자 예상과 다른 움직임들이 있어 보완해야 합니다.
  - dataset을 손수 만들다 보니 비효율적이었습니다. 조금 더 자동화할 수 있는 방법을 강구해야 합니다.
  - fetch해온 데이터를 캐싱(caching)해서 fetch 비용을 줄여야 합니다.

- 더 공부할 점
  - firebase를 조금 더 이해해야 합니다.
  - sass를 좀 더 sass처럼 써야 합니다. 

- [데모](https://sangwon21.github.io/amazon-ui/client/slider.html)

sangwon21 committed:
feat: 클라이언트 요청에 응답할 데이터 집합 생성

why: 클라이언트에서 요청이 들어오면 이에 알맞은 응답을 하기 위해,
데이터셋을 만들었다. 데이터는 영문학 고전들의 제목으로 이루어져 있다.

sangwon21 committed:
feat: nodemon 설치

why: 서버 코드 변경 시, 변경된 서버 코드를 반영하기 위해서 계속 서버를 내렸다가 올려야 했다. 이러한 번거로운 작업을, nodemon을 이용하면 피할 수 있어서 설치하게 되었다.

sangwon21 committed:
feat: search api 작성

why: 클라이언트에서 특정 문자열을 기준으로, 그 특정 문자열로 시작하는
책 제목들을 배열에 담아 보내주는, search api를 작성했다. 특히, 문자열은
대소문자를 구별하기 때문에, 대소문자 처리를 해주고 데이터를 처리하도록
했다.

sangwon21 committed:
feat: search-bar HTML 및 CSS 작업

why: search-bar HTML과 CSS 작업을 완료했다.

sangwon21 committed:
fix: CSS 수정

why: 단어 사이 간격을 늘리는 CSS를 적용했다. 또한 input창에 입력하는
단어와 줄을 맞추기 위하여 CSS를 수정했다.

sangwon21 committed:
feat: Data - UI Sync를 위한 Model 객체

why: Data와 UI 간 동기화를 위하여, 자바스크립트에서 제공하는 Proxy라는
객체를 이용하기로 했다. Proxy 객체를 이용하여 데이터가 'set'되면 해당
데이터에 맞는 UI가 렌더링되도록 만들었다. 데이터가 바뀌면, UI도 바뀐다.

sangwon21 committed:
feat: 서버 URL 추가

why: 요청을 보낼 서버 URL을 추가했다. 해당 주소로 요청을 보내게 되면,
응답으로, 요청 문자열 기준으로 정렬된 배열을 준다. 이를 클라이언트에서
렌더링하면 된다.

sangwon21 committed:
fix: slider 데이터 요청 URL 수정

why: 새로이 search API가 만들어지면서, slider UI를 제작하는 데에 필요한
카드 데이터를 얻어오는 변수값이 변했다. 이에 변한 변수값에 맞게 코드를
수정했다.

sangwon21 committed:
feat: 자동 완성 기능 추가

why: 자동 완성 기능을 부분적으로 구현했다. search-bar에서 input 이벤트로
해당 글자를 받으면, 그 해당 글자를 가지고, API 요청을 보낸다. 보내고
나서, 데이터를 Model 객체에 set해주면, 미리 등록해둔 콜백함수가 불리게
되는데, render함수가 최종적으로 불리게끔 설정해두었다. 즉, input창에
입력되는 값에 따라 보여지는 데이터가 달라지게끔 만들었다.

sangwon21 committed:
chore: server 코드 console.log 삭제 및 book_list.json 데이터 추가

sangwon21 committed:
rename: book_list 이름 교체

why: book_list json파일에 책 제목뿐만 아니라 영화 제목도 포함되게 되어
이름을 교체하였다.

sangwon21 committed:
fix: 빈 문자열일 때, fetch 보내는 오류 수정

why: 빈 문자열일 때, fetch를 보내, 서버와 통신을 했었다. 그 결과, cors
관련 오류가 계속해서 발생했다. 빈 문자열일 때, 좀 더 적절한 행동은
그려졌던 검색창이 사라지는 것이다. 그래서 fetch를 보내는 대신 검색했던
검색창이 사라지게끔 했다.

sangwon21 committed:
chore: console.log 삭제

sangwon21 committed:
feat: fetch 요청에 디바운싱 적용

why: input box에 input 이벤트가 발생할 때마다 fetch 요청을 보낸다.
하지만 이는 적절하지 못한 행동이다. 왜냐하면 한글 같은 경우,
'지키다'라는 단어 중간에 '짘'이라는 단어가 만들어 지는데, 이런 글자로
시작하는 단어는 없기 때문에 이 단어로 요청을 보내는 건, 네트워크
낭비이기 때문이다. 따라서 이런 행동을 제약하기 위하여, 사용자가 문장을
끝마쳤을 것이라고 생각되는 시간 이후로 요청을 보내겠끔 설정했다.

sangwon21 committed:
fix: 아마존 로고 추가 및 배경색 교체

why: 아마존 로고를 추가하기 위해, HTML과 CSS를 수정했다.

sangwon21 committed:
feat: 입력한 단어만 강조(하이라이팅)하는 기능 추가

why: 사용자가 입력한 단어만 다른 색깔로 하이라이팅하는 효과를 추가했다.
이를 통해 사용자는 자신이 입력한 단어와 연결되는 단어들을 무엇인지
빠르게 파악할 수 있다.

sangwon21 committed:
feat: 위 아래 방향키 이벤트 추가

why: 위 아래 방향키를 입력했을 때, 밑에 있는 단어들에 커서가 이동하게끔
이벤트를 추가했다. 엄밀히 말하면, 커서가 이동하는 게 아니라, CSS
클래스의 대상이 계속 해서 바뀐다. 일단, 이 방법이 제일 적절해 보여서, 이
방법으로 진행했다. 추후에 좀 더 좋은 방법이 생각나면 바꿀 예정이다.

sangwon21 committed:
feat: 화면을 선택하거나 버튼을 선택했을 때, 단어창이 사라지는 이벤트
추가

why: body와 button에 이벤트를 각각 추가했다. 사용자 경험을 고려했을 때,
단어창이 아닌 영역을 클릭하면, 단어창이 사라지는 방법이 맞는 것 같다.

sangwon21 committed:
feat: 커서 이동 후, enter 입력 시 input value 변경 이벤트 작성

why: 사용자가 위 아래 버튼으로 커서를 조종 후, enter를 입력 시 input
값을 변경했다. 사용자 경험 측면에서, enter를 입력하면, input value 값을
변경하는 UI 동작이 적절하다.

sangwon21 committed:
fix: enter 이벤트 이후 동작 수정

why: 사용자 경험 측면에서, enter 이후, 이미 등장했던 단어 목록창을 닫는
게 더욱 익숙하고 적절해서, enter 입력 이후,  목록창을 닫게끔 코드를
수정했다.

sangwon21 committed:
fix: 스크롤바, 하이라이트 색상, input value 등 CSS 개선

why: CSS가 통일성이 부족하여, 웹 페이지가 따로 노는 느낌을 주었다. 이를
해결하기 위하여 노란색 계열과 주황색 계열을 합친 색상으로 전체적인 색깔
틀을 잡았다. 색깔이 들어가는 곳에 일관적인 색상을 써서, 전체적으로
통일감을 주게 되었다.

sangwon21 committed:
feat: 스크롤바 상승 하강시 이동 이벤트 추가

why: 스크롤바를 조작하는 동작에서 스크롤바가 움직이지 않는다면, 사용자가
가리키는 커서와 실제 보고 있는 화면이 일치하지 않는 오류가 발생한다.
이를 해결하기 위하여, down 키를 누르면 스크롤이 내려가고, up 키를 누르면
스크롤이 올라가게끔 만들었다.

sangwon21 committed:
feat: firebase에 서버 코드 배포

why: local에서 서버를 올리면 github에 전체 코드를  올렸을 때, 요청을 제대로 보낼 수가
없다. 왜냐하면 local서버는 github에 올라가지 않기 때문이다. 그래서
클라이언트 입장에서 제대로 된 응답을 받지 못하고 결국 아무 것도 화면에
그릴 수가 없게 된다. 이에 firebase를 이용하여 서버를 배포하기로
결정했다. 일단 tutorial을 따라하면서 배포까지만 되게끔 했다. 추후에
다양한 firebase의 기능을 활용해볼 생각이다.

sangwon21 committed:
feat: 로컬 환경 주소를 서버 환경 주소로 변경

why: 서버가 firebase에서 배포됨에 따라, 로컬 환경 주소를 사용했던
부분들을 모두 firebase에 배포된 주소로 바꿨다.

sangwon21 committed:
refactor: 클라이언트 코드 중 불명확한 변수명 수정 및 상수들 변수화

why: 다른 사람이 읽었을 때, 의미를 파악하기 힘든 불명확한 변수명이나
상수들을 의미 파악이 잘 되도록, 구체적인 변수명으로 교체했다.

sangwon21 committed:
fix: search list 박스를 클릭했을 때, 박스가 닫히는 오류 해결

why: 웹 페이지 내에 어디든 클릭하면, search list 박스가 닫히게끔 했다.
이 이벤트를 document에 달아놓은 결과, search list 박스를 클릭해도 닫히게
되었다. 그래서 search list에 해당하는 클래스들을 검사한 후, 그 결과에
따라 분기 처리했다.

sangwon21 committed:
feat: 검색 결과 단어 클릭 시, input 박스에 단어 추가 이벤트 작성

why: 검색 결과 단어를 클릭하면 input 박스에 해당 단어가 추가되는 UI가
사용자 경험 측면에서 익숙하고 당연하게 생각되는 작동 방법이다. 이에 검색
결과 박스에서 단어를 선택하면 박스가 닫히면서 해당 단어를 추가하는
방식으로 이벤트를 구현했다.

crongro commented on client/src/main.js:
```diff
@@ -9,6 +13,8 @@ const addEvents = () => {
 
 const renderHTML = async () => {
+    const searchList = new SearchList();
+    const searchBar = new SearchBar(searchList);
```
main에서 같이 초기화 하는 거 좋네요.

crongro commented on client/src/util/constants.js:
```diff
@@ -1,8 +1,27 @@
-    }}
-}
\ No newline at end of file
+    DEV: {
```
굿굿 
현업에서는 production mode 뿐아니라 beta mode 등 다양한 단계로 개발하죠.

sangwon21 replied:
노드 환경에서는 process.env로 현재 개발하는 환경을 구별할 수 있을 것 같은데 브라우저에서는 그 환경을 어떻게 구별할 수 있을까요?

crongro commented on client/src/view/search_bar/model.js:
```diff
@@ -0,0 +1,21 @@
+class Model {
+    constructor(callback) {
+        const proxy = new Proxy(this, {
```
Proxy!!! 패턴을 구현한 셈이네요. 
객체모니터링하다가 어떤 변경을 해주는 장점이 있죠.
🥇

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+    ONE_SCROLL_UNIT,
+    SCROLL_TOP_START,
+    SCROLL_BOTTOM_END
+} from "../../util/constants.js";
```
너무 많으면 하나의 객체로 받아서 처리해도 될 듯.

sangwon21 replied:
네 비슷한 성질끼리 객체로 묶어서 export, import해보겠습니다.

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+        this.hitList = null;
+    }
+
+    // 참고: https://www.zerocho.com/category/JavaScript/post/59a8e9cb15ac0000182794fa
```
참고하는 거 좋고요. 
잘 이해해서 수정할 수 있으면 됨.

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+
+    // 참고: https://www.zerocho.com/category/JavaScript/post/59a8e9cb15ac0000182794fa
+    inputEventListener(event) {
+        const targetString = event.target.value;
```
destructuring문법도 활용하세요.

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+            clearTimeout(this.timer);
+        }
+
+        if (targetString === "") {
```
targetString이 공백인경우에도 this.timer 체크를 해야하는건데 맞죠? 
코드 순서가 그렇게 되어 있길래 확인차 질문.

sangwon21 replied:
네 공백인 경우라면, 존재하는 setTimeout을 없애야 한다고 생각했습니다.

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+            fetch(URL.PROD.API_SERVER_SEARCH.ADDRESS + targetString)
+                .then(res => res.json())
+                .then(titles => this.searchList.setTargetTitle(targetString, titles));
+        }.bind(this), 200);
```
arrow function을 쓰면 bind가 필요 없을까요? 
왜그렇죠?

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+        }.bind(this), 200);
+    }
+
+    isArrowKeyCode(keyCode) {
```
이런함수 좋음.

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    isSearchListArea(target) {
+        const inspectionClassList = ["hitlist-wrapper", "search-list-words", "search-list-word", "target-word", "rest-word"];
```
search_bar가 범용적으로 동작하려고 한다면, 
이런코드는 좀 아쉽죠. style 에 쓰이는 class이름이 하드코딩되어 있어서요.

sangwon21 replied:
상수화 시키겠습니다...

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+    }
+
+    handleKeyMovement(arrowDirection) {
+        if (this.isHitListOn() === false) {
```
명확한 false비교를 하고 싶으면 이럴 수 있지만 보통은 
if(this.isHitListOn()) 이라고 표현하긴 하죠.

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+        }
+        const { target } = event;
+
+        if (target.classList.contains("search-list-words") || target.classList.contains("hitlist-wrapper")) {
```
조건문을, 배열의 indexOf 를 사용해서 수정해보세요.

sangwon21 replied:
이 부분도 util 함수에 추가할 수 있을 것 같습니다.

crongro commented on client/src/view/search_bar/search_bar.js:
```diff
@@ -0,0 +1,249 @@
+        const inputBox = $("#search-bar-input");
+
+        if (target.classList.contains("search-list-word")) {
+            if (target.innerHTML === "No Results Matched") {
```
이건 더 위로 올려서 빨리 체크해서 return 시키면 어때요?

sangwon21 replied:
빨리 return해도 괜찮을 것 같습니다.

crongro commented on client/src/view/search_bar/search_bar.scss:
```diff
@@ -0,0 +1,186 @@
+        font-size: 10px;
+
+        select {
+            -webkit-appearance: none;
```
prefix 를 요즘은 도구가빌드할때 추가해주기도 해요.
하지만 이렇게 직접 추가해보는 것도 처음엔 좋죠.

crongro commented on client/src/view/search_bar/search_list.js:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    initializeHTML() {
+        $('.hitlist-wrapper').style.display = "block";
```
이런 코드는 어떻게 만들까요? 
```
$('hit-wrapper').show().html('').setScrollTop(SCROLL_TOP_START);
```

sangwon21 replied:
좋은 유틸 함수 거리를 제공해주셔서 감사합니다.

crongro reviewed (changes requested):
리뷰남겨두고 PR은 request change 상태로 둘테니, 
개선해보세요.

잘 구현중입니다. 본인을 믿고 계속 고고!

sangwon21 committed:
feat: CSS 스타일 다루는 util 함수 추가

why: 일일이 DOM 조작을 하면서 CSS를 다루는 건, 성능적인 측면에서도,
그리고 코드 가독성 측면에서도, 별로 좋지 못하다. 코드는 코드대로
길어지고, 성능은 성능대로 계속해서 해당 DOM을 찾아야하기 떄문에 좋지
못하다. 결정적으로 무엇을 하는 행동인지 명시적이지 못한 경우가 많아서
이해하기 힘들다. 이에 이런 행동들을 한 번 감싸는 방법으로 util함수를
만들었다.

sangwon21 committed:
refactor: Element util 함수 이용

why: 기존의 코드에서는 비효율적으로 계속해서 DOM을 탐색해서 찾은 다음,
정해진 행동들을 하고 있었다. 하지만 지금처럼 Element에 prototype을 달아,
이 함수를 호출한 객체를 반환하는 방식으로 chaining을 이어나간다면,
DOM을 탐색할 필요도 없고, 가독성도 개선되는 측면이 있어서 활용하게
되었다.

sangwon21 committed:
refactor: destructuring 활용, Element hide 함수 수정 및 boolean 표현식
간소화

why: Element hide의 sytle이 none이 되어야 하는데 hidden으로 잘못 적혀
있어서 바꿨다. destructuring을 활용하면 변수명 창작의 고통해서
해방되어서 사용했다. boolean 비교를 할 때, true나 false를 명시적으로
적어야지 가독성이 높은 게 아니어서 해당 코드를 삭제했다.

sangwon21 committed:
refactor: 배열에서 특정 원소가 존재하는지 확인하는 함수를 prototype으로
생성

why: 배열에서 특정 원소가 존재하는지 확인하는 작업을 계속해서 했다. 특정 DOM의 style에 특정 class가 있는지 계속 확인했다. indexOf를 통해 -1 여부를 확인하는 방식도 있지만, 계속해서 -1인지 아닌지 여부를 파악하는 일이 번거로웠다. 그래서 Array의 prototype으로 이런 일을 대신해줄 수 있는 함수를 util에 작성하게 되었다.

sangwon21 committed:
refactor: 상수를 서로 연관 있는 것들 끼리 묶음

why: 상수들을 하나씩 쓰다 보딘, 한 파일에 import을 해야되는 분량이 많이
늘어났다. 더불어, 응집력이 약해 보기 좋지 않았다. 그래서 서로 비슷한
성질의 상수들끼리 객체화 시켜서 export하기로 결정했다.

sangwon21 committed:
refactor: Search-Bar와 Search-List에 쓰는 CSS 클래스 enum화

why: 두 UI 컴포넌트에서 쓰는 CSS 클래스를 enum처럼 쓰게 했다. CSS
클래스가 기존에는 string으로 박혀있어서 변경사항에 취약했다. 그래서 이를
해결하고자 enum과 같은 방식을 활용하기로 결정했다.

sangwon21 commented:
## 고치거나 바꾼 부분
- DOM 조작 중에 유틸 함수로 만들 수 있는 부분이 있어서 유틸 함수로 만들었습니다.
- destructuring 문법을 활용했습니다.
- 의미없는 스트링 상수들을 모아 enum화 시켰습니다.
- 서로 독립적으로 존재하던 상수들을 묶어 객체로 만들어서, 조금 더 명확한 의미를 찾을 수 있게 하였습니다. 

## 더 해볼 부분
- 로컬 서버를 이용해서 개발할 때, 고려하지 않았던 문제점, 응답 속도를 해결하는 방법을 찾아야 합니다.
- 특히, 데모 주소에 처음 들어갔을 때, 검색창이 보이지 않고 새로고침을 해야만 보이는 오류가 있는데 이를 해결해야 합니다.
- 동작 중에 사용자 예상과 다른 움직임들이 있어 보완해야 합니다.

crongro commented on client/src/util/constants.js:
```diff
@@ -17,11 +17,27 @@ export const URL = {
-export const DIRECTION_UP = -1;
-export const DIRECTION_DOWN = 1;
\ No newline at end of file
+export const KEY_CODE = {
```
객체활용 좋네요.

crongro commented on client/src/util/util.js:
```diff
@@ -1,3 +1,26 @@
 export const $ = target => document.querySelector(target);
 export const $$ = target => document.querySelectorAll(target);
 
+Element.prototype.show = function () {
```
Native 확장을 알아냈군요!
이렇게 사용하면 아주 개발할때 편리할 거 같네요.

그런데 단점이 한가지 있을꺼 같은데?  뭘까요?

crongro reviewed (approved):
좋아요~ 잘 수정해나가고 계심.
